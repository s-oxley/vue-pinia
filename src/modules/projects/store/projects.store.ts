import { defineStore } from 'pinia';
import { computed, ref } from 'vue';
import type { Project, Task } from '../interfaces/project.interface';
import { useLocalStorage } from '@vueuse/core';

export const useProjectStore = defineStore('projects', () => {
  // state
  const projects = ref(useLocalStorage<Project[]>('projects', []));

  // getters
  // const getProjects = computed(() => [...projects.value]);

  const getProjects = computed(() => projects.value);
  //lo que se agrega al array desde el getProjects queda reactivo.

  // actions
  function addProject(name: string, version: number, description: string, tasks: Task[]): void {
    const newProject: Project = {
      name,
      version,
      description,
      tasks,
    };
    projects.value.push(newProject);
  }

  return { getProjects, addProject };
});
