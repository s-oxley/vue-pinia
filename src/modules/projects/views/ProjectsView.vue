<template>
  <div class="overflow-x-auto">
    <table class="table">
      <!-- head -->
      <thead>
        <tr>
          <th></th>
          <th>Nombre</th>
          <th>Trabajo</th>
          <th>Color favorito</th>
        </tr>
      </thead>
      <tbody>
        <!-- row 1 -->
        <tr>
          <th>1</th>
          <td>Santiago Oxley Heimen</td>
          <td>Product Owner</td>
          <td>Blue</td>
        </tr>
        <!-- row 2 -->
        <tr class="hover:bg-base-300">
          <th>2</th>
          <td>Javier Molina</td>
          <td>Technical Lead</td>
          <td>Purple</td>
        </tr>
        <!-- row 3 -->
        <tr>
          <th>3</th>
          <td>Brice Swyre</td>
          <td>Account Manager</td>
          <td>Red</td>
        </tr>
        <!-- row 4 -->
        <tr class="hover:bg-base-300">
          <th>4</th>
          <td>Paco Syner</td>
          <td>Developer</td>
          <td>Orange</td>
        </tr>
      </tbody>
    </table>
    <FabButton position="bottom-right" @click="modalOpen = true">
      <AddFolderIcon />
    </FabButton>
    <InputModal
      :open="modalOpen"
      @close="
        () => {
          console.log('close modal');
          modalOpen = false;
        }
      "
      @submit="
        (event) => {
          console.log(event);
          modalOpen = false;
        }
      "
    />
  </div>
</template>

<script setup lang="ts">
import FabButton from '@/modules/commons/components/FabButton.vue';
import InputModal from '@/modules/commons/components/InputModal.vue';
import AddFolderIcon from '@/modules/commons/icons/AddFolderIcon.vue';

import { ref } from 'vue';

const modalOpen = ref(false);
</script>
