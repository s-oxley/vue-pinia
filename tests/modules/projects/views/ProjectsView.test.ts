import ProjectsView from '@/modules/projects/views/ProjectsView.vue';
import { createTestingPinia } from '@pinia/testing';
import { mount } from '@vue/test-utils';
import { projectsFake } from '../../../mockData/projectsFake';
import { useProjectStore } from '@/modules/projects/store/projects.store';
// import InputModal from '@/modules/commons/components/InputModal.vue';

describe('<ProjectsView />', () => {
  const wrapper = mount(ProjectsView, {
    global: {
      plugins: [createTestingPinia()],
      stubs: ['router-link'],
    },
  });

  const store = useProjectStore();

  beforeEach(() => {
    store.$patch(() => {
      store.projects = projectsFake;
    });
  });

  test('should render projects in view', () => {
    const trs = wrapper.findAll('tbody tr');
    expect(trs.length).toBe(4);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should call addProject method on modal', () => {
    const inputModalWrapper = wrapper.findComponent({ name: 'InputModal' });
    // console.log(inputModalWrapper);
    const projectNameTest = 'Nuevo-Projecto-Test';

    inputModalWrapper.vm.$emit('submit', {
      projectName: projectNameTest,
      projectVersion: 3.4,
      projectDescription: 'N/A',
    });
    expect(store.addProject).toHaveBeenLastCalledWith(projectNameTest, 3.4, 'N/A', []);
  });
});
