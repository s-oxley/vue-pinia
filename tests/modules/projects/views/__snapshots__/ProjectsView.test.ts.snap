// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProjectsView /> > should render projects in view 1`] = `
"<div class="overflow-x-auto">
  <table class="table">
    <!-- head -->
    <thead>
      <tr>
        <th>Version</th>
        <th>Nombre</th>
        <th>Descripción</th>
        <th>Tareas</th>
        <th>Progreso</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>2.5</th>
        <td>Casupá</td>
        <td>Proyecto de Casupá</td>
        <td>3</td>
        <td><progress class="progress progress-secondary w-56" min="0" max="3" value="3"></progress></td>
      </tr>
      <tr>
        <th>3.5</th>
        <td>Uruguay</td>
        <td>Proyecto de la R.O.U.</td>
        <td>0</td>
        <td><progress class="progress progress-secondary w-56" min="0" value="0"></progress></td>
      </tr>
      <tr>
        <th>1.5</th>
        <td>Brasil</td>
        <td>Proyecto de la República del Brasil</td>
        <td>0</td>
        <td><progress class="progress progress-secondary w-56" min="0" value="0"></progress></td>
      </tr>
      <tr>
        <th>2.9</th>
        <td>Rio de los pájaros</td>
        <td>Proyecto sobre el río Uruguay</td>
        <td>0</td>
        <td><progress class="progress progress-secondary w-56" min="0" value="0"></progress></td>
      </tr>
    </tbody>
  </table><button data-v-6a7267b0="" class="btn btn-circle btn-secondary fixed bottom-right"><svg width="24px" height="24px" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
      <g id="SVGRepo_bgCarrier" stroke-width="0"></g>
      <g id="SVGRepo_tracerCarrier" stroke-linecap="round" stroke-linejoin="round"></g>
      <g id="SVGRepo_iconCarrier">
        <path d="M9 13H15M12 10V16M12.0627 6.06274L11.9373 5.93726C11.5914 5.59135 11.4184 5.4184 11.2166 5.29472C11.0376 5.18506 10.8425 5.10425 10.6385 5.05526C10.4083 5 10.1637 5 9.67452 5H6.2C5.0799 5 4.51984 5 4.09202 5.21799C3.71569 5.40973 3.40973 5.71569 3.21799 6.09202C3 6.51984 3 7.07989 3 8.2V15.8C3 16.9201 3 17.4802 3.21799 17.908C3.40973 18.2843 3.71569 18.5903 4.09202 18.782C4.51984 19 5.07989 19 6.2 19H17.8C18.9201 19 19.4802 19 19.908 18.782C20.2843 18.5903 20.5903 18.2843 20.782 17.908C21 17.4802 21 16.9201 21 15.8V10.2C21 9.0799 21 8.51984 20.782 8.09202C20.5903 7.71569 20.2843 7.40973 19.908 7.21799C19.4802 7 18.9201 7 17.8 7H14.3255C13.8363 7 13.5917 7 13.3615 6.94474C13.1575 6.89575 12.9624 6.81494 12.7834 6.70528C12.5816 6.5816 12.4086 6.40865 12.0627 6.06274Z" stroke="#fefefe" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path>
      </g>
    </svg></button>
  <!-- <FabButton position="bottom-left" @click="customModalOpen = true">
      <ClipIcon />
    </FabButton> -->
  <dialog class="modal">
    <div class="modal-box">
      <h3 class="text-lg font-bold">Nuevo proyecto</h3>
      <p class="py-4"> Completar los campos para crear un nuevo proyecto. Y presionar 'Aceptar' para crear el proyecto. Para cerrar el modal presionar 'Cerrar' </p>
      <div class="modal-action flex flex-col">
        <form method="dialog">
          <!-- @submit.prevent="$emit('submit', projectName)"> --><input type="text" placeholder="Nombre del proyecto" class="input input-bordered w-full flex-1 input-primary"><input type="number" placeholder="Versión del proyecto" class="input input-bordered w-full flex-1 input-primary mt-4"><input type="text" placeholder="Descripción del proyecto" class="input input-bordered w-full flex-1 input-primary mt-4">
          <div class="flex justify-end mt-8"><button class="btn mr-4"> Cerrar </button><!-- <button type="submit" class="btn btn-primary"> --><button type="button" class="btn btn-primary"> Aceptar </button></div>
        </form>
      </div>
    </div>
    <!-- <div
      v-if="props.open"
      class="modal-backdrop fixed top-0 left-0 z-10 bg-black opacity-40 w-screen h-screen"
    /> -->
  </dialog>
  <dialog class="modal">
    <div class="modal-box z-5">
      <div class="border-b border-blue-500">
        <h3 class="text-lg font-bold">Hola!</h3>
        <p class="py-4">Ingresa el nombre de la persona. Si quiere cerrar presione 'Cerrar'.</p>
      </div>
      <div class="modal-action flex flex-col">
        <form method="dialog "><input type="text" placeholder="Nombre" class="input input-bordered w-full flex-1 input-primary">
          <div class="flex justify-end mt-8"><button class="btn">Cerrar</button></div>
        </form>
      </div>
    </div>
    <!-- <div
      v-if="props.open"
      class="modal-backdrop fixed top-0 left-0 z-10 bg-black opacity-40 w-screen h-screen"
    /> -->
  </dialog>
</div>"
`;
